File: client-ts/integration.mjs
/**
 * Integration test for the SODP TypeScript client.
 * Run: node integration.mjs
 */
import { SodpClient } from "./dist/index.js";
import WebSocket from "ws";
import { spawn } from "child_process";
import { createHmac } from "crypto";

function makeJwt(sub, secret, ttl = 3600) {
  const h = Buffer.from(JSON.stringify({ alg: "HS256", typ: "JWT" })).toString("base64url");
  const p = Buffer.from(JSON.stringify({ sub, exp: Math.floor(Date.now() / 1000) + ttl })).toString("base64url");
  const s = createHmac("sha256", secret).update(`${h}.${p}`).digest("base64url");
  return `${h}.${p}.${s}`;
}

const sleep = ms => new Promise(r => setTimeout(r, ms));

const PORT   = 7788;
const SECRET = "integ-test-secret";

let passed = 0, failed = 0;
function ok(label) { console.log(`  ✓ ${label}`); passed++; }
function fail(label, reason) { console.error(`  ✗ ${label}: ${reason}`); failed++; }
async function assert(label, fn) {
  try { await fn(); ok(label); }
  catch (e) { fail(label, e.message); }
}

console.log("Starting SODP server on port", PORT, "...");
const proc = spawn("../target/debug/sodp-server", [`0.0.0.0:${PORT}`], {
  env: { ...process.env, RUST_LOG: "error", SODP_JWT_SECRET: SECRET },
  stdio: "pipe",
});
proc.stderr.on("data", d => process.stderr.write(d));
await sleep(400);

function newClient(sub = "user") {
  return new SodpClient(`ws://localhost:${PORT}`, { WebSocket, token: makeJwt(sub, SECRET) });
}

// ── Test 1: StateRef watch + set ─────────────────────────────────────────────
console.log("\n── Test 1: StateRef watch + set ─────────────────────────────────");
await (async () => {
  const client = newClient("t1");
  await client.ready;

  const player = client.state("integ.player");
  const events = [];
  player.watch((v, meta) => events.push({ v, meta }));
  await sleep(100); // STATE_INIT

  await player.set({ name: "Alice", health: 100, position: { x: 0, y: 0 } });
  await sleep(100);

  await assert("STATE_INIT fires with initialized=false for new key", () => {
    if (events[0]?.meta?.initialized !== false) throw new Error(JSON.stringify(events[0]?.meta));
  });
  await assert("set() fires with full value", () => {
    if (events[1]?.v?.name !== "Alice") throw new Error(JSON.stringify(events[1]?.v));
  });
  client.close();
})();

await sleep(100);

// ── Test 2: Deep patch — only named fields updated ────────────────────────────
console.log("\n── Test 2: Deep patch — only named fields updated ───────────────");
await (async () => {
  const client = newClient("t2");
  await client.ready;

  const player = client.state("integ.patch");
  player.watch(() => {});
  await player.set({ name: "Bob", health: 100, position: { x: 0, y: 0 } });
  await sleep(100);

  // patch only health — name and position must survive
  await player.patch({ health: 80 });
  await sleep(100);

  await assert("patch preserves unmentioned fields", () => {
    const snap = player.get();
    if (snap?.name !== "Bob")        throw new Error(`name: ${snap?.name}`);
    if (snap?.health !== 80)         throw new Error(`health: ${snap?.health}`);
    if (snap?.position?.x !== 0)     throw new Error(`position.x: ${snap?.position?.x}`);
  });

  // deep patch — update only position.x, y must survive
  await player.patch({ position: { x: 99 } });
  await sleep(100);

  await assert("deep patch preserves sibling nested fields (position.y)", () => {
    const snap = player.get();
    if (snap?.position?.x !== 99) throw new Error(`x: ${snap?.position?.x}`);
    if (snap?.position?.y !== 0)  throw new Error(`y: ${snap?.position?.y}`);
  });

  client.close();
})();

await sleep(100);

// ── Test 3: setIn — atomic nested field update ───────────────────────────────
console.log("\n── Test 3: setIn — atomic nested field update ───────────────────");
await (async () => {
  const client = newClient("t3");
  await client.ready;

  const player = client.state("integ.setin");
  player.watch(() => {});
  await player.set({ name: "Carol", health: 100, position: { x: 0, y: 0 } });
  await sleep(100);

  await player.setIn("/position/x", 42);
  await sleep(100);

  await assert("setIn updates only the target field", () => {
    const snap = player.get();
    if (snap?.position?.x !== 42) throw new Error(`x: ${snap?.position?.x}`);
    if (snap?.position?.y !== 0)  throw new Error(`y: ${snap?.position?.y}`);
    if (snap?.name !== "Carol")   throw new Error(`name: ${snap?.name}`);
  });

  // setIn on a deeply nested path that doesn't exist yet
  await player.setIn("/stats/kills", 5);
  await sleep(100);

  await assert("setIn creates intermediate objects", () => {
    const snap = player.get();
    if (snap?.stats?.kills !== 5) throw new Error(JSON.stringify(snap?.stats));
  });

  client.close();
})();

await sleep(100);

// ── Test 4: delete() — true removal ─────────────────────────────────────────
console.log("\n── Test 4: delete() — true removal ──────────────────────────────");
await (async () => {
  const client = newClient("t4");
  await client.ready;

  const player = client.state("integ.delete");
  const events = [];
  player.watch((v, meta) => events.push({ v, meta }));

  await player.set({ name: "Dave" });
  await sleep(100);

  events.length = 0;
  await player.delete();
  await sleep(100);

  await assert("delete fires callback with null", () => {
    if (events[0]?.v !== null) throw new Error(`v: ${JSON.stringify(events[0]?.v)}`);
  });
  await assert("delete sets initialized=false", () => {
    if (events[0]?.meta?.initialized !== false) throw new Error(`initialized: ${events[0]?.meta?.initialized}`);
  });
  await assert("get() returns null after delete", () => {
    if (player.get() !== null) throw new Error(JSON.stringify(player.get()));
  });

  client.close();
})();

await sleep(100);

// ── Test 5: fanout across two clients ────────────────────────────────────────
console.log("\n── Test 5: fanout across two clients ─────────────────────────────");
await (async () => {
  const c1 = newClient("t5a"), c2 = newClient("t5b");
  await Promise.all([c1.ready, c2.ready]);

  const s1 = c1.state("integ.fanout2"), s2 = c2.state("integ.fanout2");
  const got2 = [];
  s2.watch(v => got2.push(v));
  await sleep(100);

  await s1.set({ score: 99 });
  await sleep(100);

  await assert("client2 receives client1's update", () => {
    const last = got2[got2.length - 1];
    if (last?.score !== 99) throw new Error(JSON.stringify(last));
  });

  c1.close(); c2.close();
})();

await sleep(100);

// ── Test 6: call() before connected is queued ────────────────────────────────
console.log("\n── Test 6: call() queued before connected ────────────────────────");
await (async () => {
  const client = newClient("t6");
  const scores = client.state("integ.queue2");
  scores.watch(() => {});

  // Fire immediately — not yet authenticated
  const p = scores.set({ queued: true });
  await p;
  await sleep(100);

  await assert("queued call executes after connect", () => {
    const snap = scores.get();
    if (snap?.queued !== true) throw new Error(JSON.stringify(snap));
  });

  client.close();
})();

await sleep(100);

// ── Test 7: ready + onConnect/onDisconnect ───────────────────────────────────
console.log("\n── Test 7: ready + onConnect/onDisconnect ────────────────────────");
await (async () => {
  let connects = 0, disconnects = 0;
  const client2 = new SodpClient(`ws://localhost:${PORT}`, {
    WebSocket,
    token: makeJwt("t7", SECRET),
    onConnect:    () => connects++,
    onDisconnect: () => disconnects++,
  });

  await client2.ready;
  await assert("ready resolves", () => {});
  await assert("onConnect fired", () => { if (connects !== 1) throw new Error(`connects=${connects}`); });

  client2.close();
  await sleep(100);
  await assert("onDisconnect fired", () => { if (disconnects !== 1) throw new Error(`disconnects=${disconnects}`); });
})();

await sleep(100);

// ── Test 8: isWatching + unsubscribe ─────────────────────────────────────────
console.log("\n── Test 8: isWatching + unsubscribe ─────────────────────────────");
await (async () => {
  const client = newClient("t8");
  await client.ready;

  const ref = client.state("integ.unsub2");
  await assert("isWatching false before watch()", () => {
    if (ref.isWatching()) throw new Error("should not be watching yet");
  });

  const unsub = ref.watch(() => {});
  await sleep(50);
  await assert("isWatching true after watch()", () => {
    if (!ref.isWatching()) throw new Error("should be watching");
  });

  unsub();
  // Note: isWatching stays true (subscription entry remains, just no callbacks)
  // This is by design — key cache is preserved even with 0 callbacks

  client.close();
})();

// ── Test 9: RESUME — delta replay after reconnect ─────────────────────────────
console.log("\n── Test 9: RESUME — delta replay after reconnect ────────────────");
await (async () => {
  // Promise that resolves the *second* time onConnect fires (= first reconnect).
  let connectCount = 0;
  let reconnectResolve;
  const reconnected = new Promise(r => { reconnectResolve = r; });

  const client = new SodpClient(`ws://localhost:${PORT}`, {
    WebSocket,
    token:          makeJwt("t9", SECRET),
    reconnect:      true,
    reconnectDelay: 100,
    onConnect: () => { if (++connectCount === 2) reconnectResolve(); },
  });
  await client.ready;

  const ref = client.state("integ.resume");
  const events = [];
  ref.watch((v, meta) => events.push({ v, meta }));
  await sleep(100); // STATE_INIT

  // Write version 1 while connected.
  await ref.set({ counter: 1 });
  await sleep(50);
  const vBefore = ref.get()?.counter;

  // Drop the connection — client will RESUME on reconnect.
  client["ws"]?.close();
  await sleep(50); // let disconnect event fire

  // Write version 2 while the client is offline.
  const c2 = newClient("t9b");
  await c2.ready;
  await c2.state("integ.resume").set({ counter: 2 });
  c2.close();

  // Wait for the client to reconnect and finish RESUME (not a fixed sleep).
  await reconnected;
  await sleep(150); // allow RESUME delta + STATE_INIT to be processed

  await assert("counter was 1 before disconnect", () => {
    if (vBefore !== 1) throw new Error(`vBefore=${vBefore}`);
  });
  await assert("after RESUME, counter reflects missed update (=2)", () => {
    const snap = ref.get();
    if (snap?.counter !== 2) throw new Error(`counter=${snap?.counter}`);
  });
  await assert("RESUME fires watch callback with missed delta", () => {
    const last = events[events.length - 1];
    if (last?.v?.counter !== 2) throw new Error(JSON.stringify(last));
  });

  client.close();
})();

// ── Test 10: UNWATCH — server stops sending DELTAs ───────────────────────────
console.log("\n── Test 10: UNWATCH — server stops sending DELTAs ──────────────");
await (async () => {
  const writer = newClient("t10w");
  const reader = newClient("t10r");
  await Promise.all([writer.ready, reader.ready]);

  const rRef = reader.state("integ.unwatch");
  const events = [];
  rRef.watch(v => events.push(v));
  await sleep(100); // STATE_INIT

  // Write 1 — reader is watching, should receive it.
  await writer.state("integ.unwatch").set({ n: 1 });
  await sleep(100);

  await assert("reader receives delta before unwatch", () => {
    const last = events[events.length - 1];
    if (last?.n !== 1) throw new Error(JSON.stringify(last));
  });

  // Unwatch — reader cancels its subscription.
  rRef.unwatch();
  await sleep(50); // give server time to process UNWATCH

  await assert("isWatching() is false after unwatch()", () => {
    if (rRef.isWatching()) throw new Error("should not be watching");
  });
  await assert("get() returns undefined after unwatch()", () => {
    if (rRef.get() !== undefined) throw new Error(`got: ${JSON.stringify(rRef.get())}`);
  });

  const countBefore = events.length;

  // Write 2 — reader is no longer subscribed, must NOT receive it.
  await writer.state("integ.unwatch").set({ n: 2 });
  await sleep(100);

  await assert("reader receives no more deltas after unwatch", () => {
    if (events.length !== countBefore) throw new Error(`got ${events.length - countBefore} unexpected event(s)`);
  });

  // Re-watching the same key after unwatch should work cleanly.
  const events2 = [];
  rRef.watch(v => events2.push(v));
  await sleep(100);

  await assert("re-watch after unwatch receives current value", () => {
    const last = events2[events2.length - 1];
    if (last?.n !== 2) throw new Error(JSON.stringify(last));
  });

  writer.close(); reader.close();
})();

await sleep(100);

// ── Summary ───────────────────────────────────────────────────────────────────
await sleep(100);
proc.kill();
console.log(`\n${"─".repeat(52)}`);
console.log(`Results: ${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: sodp-py/integration.py
"""
Integration test for the SODP Python client.
Run: python integration.py
"""

import asyncio
import base64
import hashlib
import hmac
import json
import os
import subprocess
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "src"))
from sodp import SodpClient

# ── JWT helper ────────────────────────────────────────────────────────────────

def make_jwt(sub: str, secret: str, ttl: int = 3600) -> str:
    def b64(data: bytes) -> str:
        return base64.urlsafe_b64encode(data).rstrip(b"=").decode()

    header  = b64(json.dumps({"alg": "HS256", "typ": "JWT"}).encode())
    payload = b64(json.dumps({"sub": sub, "exp": int(time.time()) + ttl}).encode())
    msg     = f"{header}.{payload}".encode()
    sig     = b64(hmac.new(secret.encode(), msg, hashlib.sha256).digest())
    return f"{header}.{payload}.{sig}"

# ── Test harness ──────────────────────────────────────────────────────────────

passed = failed = 0

def ok(label: str) -> None:
    global passed
    print(f"  ✓ {label}")
    passed += 1

def fail(label: str, reason: str) -> None:
    global failed
    print(f"  ✗ {label}: {reason}", file=sys.stderr)
    failed += 1

async def assert_test(label: str, fn) -> None:
    try:
        result = fn()
        if asyncio.iscoroutine(result):
            await result
        ok(label)
    except Exception as e:
        fail(label, str(e))

PORT   = 7789
SECRET = "integ-test-secret-py"

def new_client(sub: str = "user") -> SodpClient:
    return SodpClient(
        f"ws://localhost:{PORT}",
        token=make_jwt(sub, SECRET),
    )

# ── Tests ─────────────────────────────────────────────────────────────────────

async def test1():
    print("\n── Test 1: StateRef watch + set ─────────────────────────────────")
    client = new_client("t1")
    await client.ready

    player = client.state("integ.player")
    events = []
    player.watch(lambda v, m: events.append((v, m)))
    await asyncio.sleep(0.1)   # STATE_INIT

    await player.set({"name": "Alice", "health": 100, "position": {"x": 0, "y": 0}})
    await asyncio.sleep(0.1)

    async def check_init():
        if events[0][1].initialized is not False:
            raise AssertionError(f"meta={events[0][1]}")
    await assert_test("STATE_INIT fires with initialized=False for new key", check_init)

    async def check_set():
        if events[1][0].get("name") != "Alice":
            raise AssertionError(f"v={events[1][0]}")
    await assert_test("set() fires with full value", check_set)

    client.close()


async def test2():
    print("\n── Test 2: Deep patch — only named fields updated ───────────────")
    client = new_client("t2")
    await client.ready

    player = client.state("integ.patch_py")
    player.watch(lambda v, m: None)
    await player.set({"name": "Bob", "health": 100, "position": {"x": 0, "y": 0}})
    await asyncio.sleep(0.1)

    await player.patch({"health": 80})
    await asyncio.sleep(0.1)

    async def check_patch():
        s = player.get()
        if s.get("name") != "Bob":   raise AssertionError(f"name={s.get('name')}")
        if s.get("health") != 80:    raise AssertionError(f"health={s.get('health')}")
        if s.get("position", {}).get("x") != 0:
            raise AssertionError(f"position.x={s.get('position',{}).get('x')}")
    await assert_test("patch preserves unmentioned fields", check_patch)

    await player.patch({"position": {"x": 99}})
    await asyncio.sleep(0.1)

    async def check_deep_patch():
        s = player.get()
        if s.get("position", {}).get("x") != 99:
            raise AssertionError(f"x={s.get('position',{}).get('x')}")
        if s.get("position", {}).get("y") != 0:
            raise AssertionError(f"y={s.get('position',{}).get('y')}")
    await assert_test("deep patch preserves sibling nested fields (position.y)", check_deep_patch)

    client.close()


async def test3():
    print("\n── Test 3: setIn — atomic nested field update ───────────────────")
    client = new_client("t3")
    await client.ready

    player = client.state("integ.setin_py")
    player.watch(lambda v, m: None)
    await player.set({"name": "Carol", "health": 100, "position": {"x": 0, "y": 0}})
    await asyncio.sleep(0.1)

    await player.set_in("/position/x", 42)
    await asyncio.sleep(0.1)

    async def check_setin():
        s = player.get()
        if s.get("position", {}).get("x") != 42:
            raise AssertionError(f"x={s.get('position',{}).get('x')}")
        if s.get("position", {}).get("y") != 0:
            raise AssertionError(f"y={s.get('position',{}).get('y')}")
        if s.get("name") != "Carol":
            raise AssertionError(f"name={s.get('name')}")
    await assert_test("setIn updates only the target field", check_setin)

    await player.set_in("/stats/kills", 5)
    await asyncio.sleep(0.1)

    async def check_setin_create():
        s = player.get()
        if s.get("stats", {}).get("kills") != 5:
            raise AssertionError(f"stats={s.get('stats')}")
    await assert_test("setIn creates intermediate objects", check_setin_create)

    client.close()


async def test4():
    print("\n── Test 4: delete() — true removal ──────────────────────────────")
    client = new_client("t4")
    await client.ready

    player = client.state("integ.delete_py")
    events = []
    player.watch(lambda v, m: events.append((v, m)))
    await player.set({"name": "Dave"})
    await asyncio.sleep(0.1)

    events.clear()
    await player.delete()
    await asyncio.sleep(0.1)

    async def check_null():
        if events[0][0] is not None:
            raise AssertionError(f"v={events[0][0]}")
    await assert_test("delete fires callback with None", check_null)

    async def check_uninit():
        if events[0][1].initialized is not False:
            raise AssertionError(f"initialized={events[0][1].initialized}")
    await assert_test("delete sets initialized=False", check_uninit)

    async def check_get():
        if player.get() is not None:
            raise AssertionError(f"get()={player.get()}")
    await assert_test("get() returns None after delete", check_get)

    client.close()


async def test5():
    print("\n── Test 5: fanout across two clients ─────────────────────────────")
    c1 = new_client("t5a")
    c2 = new_client("t5b")
    await asyncio.gather(c1.ready, c2.ready)

    s1 = c1.state("integ.fanout_py")
    s2 = c2.state("integ.fanout_py")
    got2 = []
    s2.watch(lambda v, m: got2.append(v))
    await asyncio.sleep(0.1)

    await s1.set({"score": 99})
    await asyncio.sleep(0.1)

    async def check_fanout():
        last = got2[-1] if got2 else None
        if not last or last.get("score") != 99:
            raise AssertionError(f"last={last}")
    await assert_test("client2 receives client1's update", check_fanout)

    c1.close()
    c2.close()


async def test6():
    print("\n── Test 6: call() queued before connected ────────────────────────")
    client = new_client("t6")
    scores = client.state("integ.queue_py")
    scores.watch(lambda v, m: None)

    # Fire immediately — not yet authenticated.
    task = asyncio.create_task(scores.set({"queued": True}))
    await task
    await asyncio.sleep(0.1)

    async def check_queued():
        s = scores.get()
        if not s or s.get("queued") is not True:
            raise AssertionError(f"snap={s}")
    await assert_test("queued call executes after connect", check_queued)

    client.close()


async def test7():
    print("\n── Test 7: ready + on_connect/on_disconnect ──────────────────────")
    connects = disconnects = 0

    def on_conn():    nonlocal connects;    connects    += 1
    def on_disc():    nonlocal disconnects; disconnects += 1

    client = SodpClient(
        f"ws://localhost:{PORT}",
        token=make_jwt("t7", SECRET),
        on_connect=on_conn,
        on_disconnect=on_disc,
    )
    await client.ready

    await assert_test("ready resolves", lambda: None)

    async def check_connect():
        if connects != 1:
            raise AssertionError(f"connects={connects}")
    await assert_test("on_connect fired", check_connect)

    client.close()
    await asyncio.sleep(0.2)

    async def check_disconnect():
        if disconnects != 1:
            raise AssertionError(f"disconnects={disconnects}")
    await assert_test("on_disconnect fired", check_disconnect)


async def test8():
    print("\n── Test 8: is_watching + unsubscribe ────────────────────────────")
    client = new_client("t8")
    await client.ready

    ref = client.state("integ.unsub_py")

    async def check_before():
        if ref.is_watching():
            raise AssertionError("should not be watching yet")
    await assert_test("is_watching False before watch()", check_before)

    unsub = ref.watch(lambda v, m: None)
    await asyncio.sleep(0.05)

    async def check_after():
        if not ref.is_watching():
            raise AssertionError("should be watching")
    await assert_test("is_watching True after watch()", check_after)

    unsub()  # removes the callback; server subscription stays alive

    client.close()


async def test9():
    print("\n── Test 9: RESUME — delta replay after reconnect ────────────────")
    connect_count = 0
    reconnected   = asyncio.Event()

    def on_conn():
        nonlocal connect_count
        connect_count += 1
        if connect_count == 2:
            reconnected.set()

    client = SodpClient(
        f"ws://localhost:{PORT}",
        token=make_jwt("t9", SECRET),
        reconnect=True,
        reconnect_delay=0.1,
        on_connect=on_conn,
    )
    await client.ready

    ref    = client.state("integ.resume_py")
    events = []
    ref.watch(lambda v, m: events.append((v, m)))
    await asyncio.sleep(0.1)   # STATE_INIT

    await ref.set({"counter": 1})
    await asyncio.sleep(0.05)
    v_before = (ref.get() or {}).get("counter")

    # Drop the connection — client will RESUME on reconnect.
    if client._ws:
        await client._ws.close()
    await asyncio.sleep(0.05)

    # Write version 2 while the client is offline.
    c2 = new_client("t9b")
    await c2.ready
    await c2.state("integ.resume_py").set({"counter": 2})
    c2.close()

    # Wait for the client to reconnect and finish RESUME.
    await asyncio.wait_for(reconnected.wait(), timeout=5.0)
    await asyncio.sleep(0.2)   # allow RESUME delta + STATE_INIT to be processed

    async def check_before():
        if v_before != 1:
            raise AssertionError(f"v_before={v_before}")
    await assert_test("counter was 1 before disconnect", check_before)

    async def check_resume():
        s = ref.get()
        if not s or s.get("counter") != 2:
            raise AssertionError(f"counter={s}")
    await assert_test("after RESUME, counter reflects missed update (=2)", check_resume)

    async def check_callback():
        last = events[-1] if events else None
        if not last or (last[0] or {}).get("counter") != 2:
            raise AssertionError(f"last={last}")
    await assert_test("RESUME fires watch callback with missed delta", check_callback)

    client.close()


async def test10():
    print("\n── Test 10: UNWATCH — server stops sending DELTAs ──────────────")
    writer = new_client("t10w")
    reader = new_client("t10r")
    await asyncio.gather(writer.ready, reader.ready)

    r_ref  = reader.state("integ.unwatch_py")
    events = []
    r_ref.watch(lambda v, m: events.append(v))
    await asyncio.sleep(0.1)

    await writer.state("integ.unwatch_py").set({"n": 1})
    await asyncio.sleep(0.1)

    async def check_before():
        last = events[-1] if events else None
        if not last or last.get("n") != 1:
            raise AssertionError(f"last={last}")
    await assert_test("reader receives delta before unwatch", check_before)

    r_ref.unwatch()
    await asyncio.sleep(0.05)

    async def check_not_watching():
        if r_ref.is_watching():
            raise AssertionError("should not be watching")
    await assert_test("is_watching() is False after unwatch()", check_not_watching)

    async def check_get_none():
        if r_ref.get() is not None:
            raise AssertionError(f"got: {r_ref.get()}")
    await assert_test("get() returns None after unwatch()", check_get_none)

    count_before = len(events)
    await writer.state("integ.unwatch_py").set({"n": 2})
    await asyncio.sleep(0.1)

    async def check_no_more():
        if len(events) != count_before:
            raise AssertionError(f"got {len(events) - count_before} unexpected event(s)")
    await assert_test("reader receives no more deltas after unwatch", check_no_more)

    # Re-watch should work cleanly.
    events2 = []
    r_ref.watch(lambda v, m: events2.append(v))
    await asyncio.sleep(0.1)

    async def check_rewatch():
        last = events2[-1] if events2 else None
        if not last or last.get("n") != 2:
            raise AssertionError(f"last={last}")
    await assert_test("re-watch after unwatch receives current value", check_rewatch)

    writer.close()
    reader.close()


# ── Main ──────────────────────────────────────────────────────────────────────

async def main():
    server_bin = os.path.join(
        os.path.dirname(__file__), "..", "target", "debug", "sodp-server"
    )
    print(f"Starting SODP server on port {PORT} …")
    proc = subprocess.Popen(
        [server_bin, f"0.0.0.0:{PORT}"],
        env={**os.environ, "RUST_LOG": "error", "SODP_JWT_SECRET": SECRET},
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )

    try:
        await asyncio.sleep(0.4)

        await test1();  await asyncio.sleep(0.1)
        await test2();  await asyncio.sleep(0.1)
        await test3();  await asyncio.sleep(0.1)
        await test4();  await asyncio.sleep(0.1)
        await test5();  await asyncio.sleep(0.1)
        await test6();  await asyncio.sleep(0.1)
        await test7();  await asyncio.sleep(0.1)
        await test8();  await asyncio.sleep(0.1)
        await test9();  await asyncio.sleep(0.1)
        await test10(); await asyncio.sleep(0.1)

    finally:
        proc.terminate()
        proc.wait()

    print(f"\n{'─' * 52}")
    print(f"Results: {passed} passed, {failed} failed")
    if failed:
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
